### PR TITLE
PYIC-790: Update the passport-back issuecredential lambda to create a sign a verifiable credential

### PR DESCRIPTION
**Pull request by Danomanic**
### What changed

Update the passport-back issuecredential lambda to create a sign a verifiable credential

### Issue tracking

- [PYI-790](https://govukverify.atlassian.net/browse/PYI-790)

### CODE REVIEW

File: lambdas/authorizationcode/src/main/java/uk/gov/di/ipv/cri/passport/authorizationcode/AuthorizationCodeHandler.java [hunk 1]
```diff
@@ -154,15 +154,15 @@ private void validateDcsResponse(DcsResponse dcsResponse)
         }
     }
 
-    private PassportGpg45Score generateGpg45Score(DcsResponse dcsResponse) {
+    private Evidence generateGpg45Score(DcsResponse dcsResponse) {
```
Comment (dannyhayes-gds) [inline lambdas/authorizationcode/src/main/java/uk/gov/di/ipv/cri/passport/authorizationcode/AuthorizationCodeHandler.java:157]: Minor thing here but why the change in class name? Seems a bit confusing now that we have "Evidence" & a "Gpg45Evidence".

File: lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java [hunk 2]
```diff
@@ -72,15 +97,44 @@ public APIGatewayProxyResponseEvent handleRequest(
...
+            return ApiGatewayResponseGenerator.proxyJoseResponse(
+                    HttpStatus.SC_OK, signedJWT.serialize());
         } catch (ParseException e) {
             LOGGER.error("Failed to parse access token");
             return ApiGatewayResponseGenerator.proxyJsonResponse(
                     e.getErrorObject().getHTTPStatusCode(), e.getErrorObject().toJSONObject());
+        } catch (JOSEException e) {
+            LOGGER.error("Failed to create VerifiableIdentityCredential");
+            return ApiGatewayResponseGenerator.proxyJsonResponse(
+                    HttpStatus.SC_INTERNAL_SERVER_ERROR, e.getLocalizedMessage());
```
Comment (dannyhayes-gds) [inline lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java:114]: Not sure if this is the ideal kind of error we'd want to return. I'm not 100% sure as the error handling in general for CRI's is a bit inconsitant and hasn't fully been figured out. But so far I think we have either returned a Nimbus ErrorObject for proper Oauth errors, or one of our ErrorResponse objects

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import java.util.UUID;
+
+public class CredentialSubject {
+
+    @JsonProperty private final Name name;
+    @JsonProperty private final String passportNumber;
+
+    @JsonProperty
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
+    private final LocalDate birthDate;
```
Comment (dannyhayes-gds) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java:17]: Does this var type now need to be the new BirthDate class

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    @JsonProperty
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
+    private final LocalDate expiryDate;
```
Comment (dannyhayes-gds) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java:21]: And same here I guess, do we need it stored in the same format?

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/NamePartType.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.di.ipv.cri.passport.library.domain;
+
+public enum NamePartType {
+    GIVEN_NAME("GivenName"),
+    FAMILY_NAME("FamilyName"),
```
Comment (dannyhayes-gds) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/NamePartType.java:5]: Don't need this comma I think?

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/NameParts.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
+package uk.gov.di.ipv.cri.passport.library.domain;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.annotation.JsonPropertyOrder;
+import uk.gov.di.ipv.cri.passport.library.annotations.ExcludeFromGeneratedCoverageReport;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({"value", "type", "validFrom", "validUntil"})
```
Comment (dannyhayes-gds) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/NameParts.java:9]: Do we need to specify the json property order?

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/VerifiableCredential.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import uk.gov.di.ipv.cri.passport.library.persistence.item.PassportCheckDao;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class VerifiableCredential {
+
+    @JsonProperty private String resourceId;
+    @JsonProperty private CredentialSubject credentialSubject;
+    @JsonProperty private Evidence evidence;
```
Comment (dannyhayes-gds) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/VerifiableCredential.java:13]: Something I'm curious about is what this but of a VC looks like for something that isn't evidence based. I.e i believe different cri's have different gpg45 score structures essentially.

Might be something to double check with george and will help us decide a good naming convention across all the CRI's

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/NameParts.java (hunk 6 above)
Comment (Danomanic) [reply lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/NameParts.java:9]: No, don't think we do!

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+        try {
+            signingInputHash = MessageDigest.getInstance("SHA-256").digest(signingInput);
+        } catch (NoSuchAlgorithmException e) {
+            throw new JOSEException(e.getMessage());
+        }
+
+        SignRequest signRequest =
+                new SignRequest()
+                        .withSigningAlgorithm(
+                                SigningAlgorithmSpec.RSASSA_PKCS1_V1_5_SHA_256.toString())
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java:57]: We're moving more towards using EC keys rather than RSA, to be more inline with what SPOT and (I believe) the KBV team are doing. This will need updating to `SigningAlgorithmSpec.ECDSA_SHA_256`.

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+                        .withMessageType(MessageType.DIGEST);
+
+        SignResult signResult = kmsClient.sign(signRequest);
+
+        return new Base64URL(b64UrlEncoder.encodeToString(signResult.getSignature().array()));
+    }
+
+    @Override
+    public Set<JWSAlgorithm> supportedJWSAlgorithms() {
+        return Set.of(RS256);
```
Comment (Wynndow) [inline lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java:69]: Also this to `ES256`

File: lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java [hunk 10]
```diff
@@ -72,15 +97,44 @@ public APIGatewayProxyResponseEvent handleRequest(
...
+            VerifiableCredential verifiableCredential) throws JOSEException {
+        Instant now = Instant.now();
+
+        // TODO: Set Subject and Issuer
+        JWTClaimsSet claimsSet =
+                new JWTClaimsSet.Builder()
+                        .claim(SUBJECT, "Subject")
+                        .claim(ISSUER, "Issuer")
+                        .claim(NOT_BEFORE, now.getEpochSecond())
+                        .claim(EXPIRATION_TIME, now.plusSeconds(600).getEpochSecond())
```
Comment (Wynndow) [inline lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java:128]: The expiration time should probably be a config value grabbed from the param store.

File: lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java [hunk 11]
```diff
@@ -72,15 +97,44 @@ public APIGatewayProxyResponseEvent handleRequest(
...
-            PassportCriResponse passportCriResponse =
-                    PassportCriResponse.fromPassportCheckDao(passportCheck);
+            VerifiableCredential verifiableCredential =
+                    VerifiableCredential.fromPassportCheckDao(passportCheck);
 
-            return ApiGatewayResponseGenerator.proxyJsonResponse(
-                    HttpStatus.SC_OK, passportCriResponse);
+            SignedJWT signedJWT = generateAndSignVerifiableCredentialJwt(verifiableCredential);
+
+            return ApiGatewayResponseGenerator.proxyJoseResponse(
```
Comment (Wynndow) [inline lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java:105]: The `proxyJoseResponse` method will set the response content type to `application/jose`. We'll need to set it to `application/jwt` when returning the signed VC or core will blow up when calling `response.getContentAsJWT()`. It might be that you can just rename this method and change the content type it it's not used elsewhere. Otherwise you might need a new method.

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java (hunk 8 above)
Comment (Danomanic) [reply lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java:57]: Nice! Updated.

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java (hunk 9 above)
Comment (Danomanic) [reply lib/src/main/java/uk/gov/di/ipv/cri/passport/library/helpers/KmsSigner.java:69]: Also done!

File: lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java (hunk 2 above)
Comment (Danomanic) [reply lambdas/issuecredential/src/main/java/uk/gov/di/ipv/cri/passport/issuecredential/IssueCredentialHandler.java]: I've fixed this to return an OAuth error status and message.

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java (hunk 3 above)
Comment (Danomanic) [reply lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java]: Done

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java (hunk 4 above)
Comment (Danomanic) [reply lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/CredentialSubject.java]: I suggest we change this once the format for expiryDate has been decided in the VC Schema.

File: lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/verifiablecredential/VerifiableCredential.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import uk.gov.di.ipv.cri.passport.library.persistence.item.PassportCheckDao;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class VerifiableCredential {
+
+    @JsonProperty private String resourceId;
```
[the remark this answers is not in the document]
Comment (Danomanic) [reply lib/src/main/java/uk/gov/di/ipv/cri/passport/library/domain/verifiablecredential/VerifiableCredential.java]: Nice spot - removed.

**Review by Wynndow — APPROVED** [review]
